gk/fib.c

```diff
@@ -0,0 +1,1326 @@
+
+out:
+	/* Initialize the fib entry for the IP prefix. */
+	ip_prefix_fib = add_prefix_fib(
+		&ip_prefix_addr, ip_prefix_len, gk_conf);
+	if (ip_prefix_fib != NULL) {
+		/* Fills up the FIB entry for the IP prefix. */
+		ip_prefix_fib->action = GK_FWD_GRANTOR;
+		rte_memcpy(&ip_prefix_fib->u,
+			&gt_fib->u, sizeof(ip_prefix_fib->u));
+
+		ip_prefix_fib->u.grantor.eth_cache->ref_cnt++;
+		gt_fib->u.grantor.next_fib->ref_cnt++;
+	} else {
+		del_fib_entry(grantor, gk_conf);
+		return NULL;
+	}
+
+	return gt_fib;
+}
+
+static struct gk_fib *
+init_drop_fib(const char *ip_prefix, struct gk_config *gk_conf)
+{
+	struct ipaddr ip_prefix_addr;
+	int ip_prefix_len;
+	struct gk_fib *ip_prefix_fib;
+
+	/* Initialize the fib entry for the IP prefix. */
+	ip_prefix_len = parse_ip_prefix(ip_prefix, &ip_prefix_addr);
+	if (ip_prefix_len < 0)
+		return NULL;
+
+	ip_prefix_fib = add_prefix_fib(
+		&ip_prefix_addr, ip_prefix_len, gk_conf);
+	if (ip_prefix_fib == NULL)
+		return NULL;
+
+	ip_prefix_fib->action = GK_DROP;
+
+	return ip_prefix_fib;
+}
+
+static int
+add_fib_entry(const char *ip_prefix, enum gk_fib_action action,
+	const char *grantor_or_gateway, struct gk_config *gk_conf)
+{
+	int ret;
+	struct gk_fib *gw_fib = NULL;
+	struct gk_fib *gt_fib = NULL;
+
+	if (ip_prefix == NULL || grantor_or_gateway == NULL)
+		return -1;
+
+	switch (action) {
+	case GK_FWD_GRANTOR: {
+
+		gt_fib = init_grantor_fib(
+			ip_prefix, grantor_or_gateway, gk_conf);
+		if (gt_fib == NULL) {
+			ret = -1;
+			goto out;
+		}
+
+		/*
+	 	 * TODO The nexthop MAC address should be initialized.
+	 	 */
+		break;
+	}
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET:
+
+		gw_fib = init_gateway_fib(ip_prefix,
+			action, grantor_or_gateway, gk_conf);
+		if (gw_fib == NULL) {
+			ret = -1;
+			goto out;
+		}
+
+		/*
+	 	 * TODO The nexthop MAC address should be initialized.
+	 	 */
+		break;
+
+	case GK_DROP: {
+
+		struct gk_fib *ip_prefix_fib =
+			init_drop_fib(ip_prefix, gk_conf);
+		if (ip_prefix_fib == NULL) {
+			ret = -1;
+			goto out;
+		}
+
+		break;
+	}
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		/* FALLTHROUGH */
+	default:
+		RTE_LOG(ERR, GATEKEEPER,
+			"Invalid fib action %u at %s\n", action, __func__);
+		ret = -1;
+		goto out;
+		break;
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"gk: add a fib entry [ip prefix = %s, action = %u, grantor_or_gateway = %s]\n",
+		ip_prefix, action, grantor_or_gateway);
+
+	ret = 0;
+out:
+	return ret;
+}
+
+int
+add_fib_entry_lock(const char *ip_prefix, enum gk_fib_action action,
+	const char *grantor_or_gateway, struct gk_config *gk_conf)
+{
+	int ret;
+
+	rte_spinlock_lock_tm(&gk_conf->lpm_tbl.lock);
+	ret = add_fib_entry(ip_prefix, action,
+		grantor_or_gateway, gk_conf);
+	rte_spinlock_unlock_tm(&gk_conf->lpm_tbl.lock);
+
+	return ret;
+}
+
+int
+del_fib_entry_lock(const char *ip_prefix, struct gk_config *gk_conf)
```

Comment:
Between the names `add_fib_entry_locked()` and `__add_fib_entry()` for the auxilary function, the name `add_fib_entry_locked()` is better.